src/opm/common/utility/numeric/calc_cellvol.cpp

```diff
@@ -0,0 +1,199 @@
+#include <opm/common/utility/numeric/calc_cellvol.hpp>
+
+
+/* 
+    Cell volume calculation based on following publication:
+   
+    D. K Pointing, Corner Point Geometry in Reservoir Simulation , 
+    ECMOR I - 1st European Conference on the Mathematics of Oil Recovery, 
+    1989 
+*/
+ 
+using namespace std;
+
+
+
+
+double C(const std::vector<double>& r,int i1,int i2,int i3){
+
+   double temp;
+
+   temp=0.0;
+   
+   if ((i1==0) && (i2==0) && (i3==0)){
+      temp=r[0];
+   }
+
+   if ((i1==1) && (i2==0) && (i3==0)){
+      temp=r[1]-r[0];
+   }
+
+   if ((i1==0) && (i2==1) && (i3==0)){
+      temp=r[2]-r[0];
+   }
+
+   if ((i1==0) && (i2==0) && (i3==1)){
+      temp=r[4]-r[0];
+   }
+
+   if ((i1==1) && (i2==1) && (i3==0)){
+      temp=r[3]+r[0]-r[2]-r[1];
+   }
+
+   if ((i1==0) && (i2==1) && (i3==1)){
+      temp=r[6]+r[0]-r[4]-r[2];
+   }
+
+   if ((i1==1) && (i2==0) && (i3==1)){
+      temp=r[5]+r[0]-r[4]-r[1];
+   }
+
+   if ((i1==1) && (i2==1) && (i3==1)){
+      temp=r[7]+r[4]+r[2]+r[1]-r[6]-r[5]-r[3]-r[0];
+   }
+
+   return temp;
+
+}
+
+double perm123(int i1,int i2,int i3){
+
+   int temp;
+   
+   if ((i1==1) && (i2==2) && (i3==3)){
+      temp=0;
+   }
+   
+   if ((i1==1) && (i2==3) && (i3==2)){
+      temp=1;
+   }
+   
+   if ((i1==2) && (i2==1) && (i3==3)){
+      temp=1;
+   }
+   
+   if ((i1==2) && (i2==3) && (i3==1)){
+      temp=2;
+   }
+   
+   if ((i1==3) && (i2==1) && (i3==2)){
+      temp=2;
+   }
+   
+   if ((i1==3) && (i2==2) && (i3==1)){
+      temp=3;
+   }
+
+   
+   return temp;
+
+}
+
+
+
+double calc_cell_volume(const std::vector<double>& X,const std::vector<double>& Y,const std::vector<double>& Z){
```

Comment:
I do not immediately understand the algorithm here, please add sufficient comments to understand the basic idea without needing to read the paper. While you add those comments, I'll look at the paper though...